include/boost/geometry/util/promote_integral.hpp

```diff
@@ -0,0 +1,276 @@
+>
+class promote_integral
+{
+private:
+    static bool const is_unsigned = boost::is_unsigned<T>::type::value;
+
+#if !defined(BOOST_GEOMETRY_NO_MULTIPRECISION_INTEGER)
+    // Define the proper check policy for the multiprecision integer
+    typedef typename boost::mpl::if_c
+        <
+            UseCheckedMultiprecisionInteger,
+            boost::integral_constant
+                <
+                    boost::multiprecision::cpp_int_check_type,
+                    boost::multiprecision::checked
+                >,
+            boost::integral_constant
+                <
+                    boost::multiprecision::cpp_int_check_type,
+                    boost::multiprecision::unchecked
+                >
+        >::type check_policy_type;
+
+    // Meta-function to get the multiprecision integer type for the
+    // given size and type (signed/unsigned)
+    template
+    <
+        unsigned int Size,
+        boost::multiprecision::cpp_integer_type IntegerType
+    >
+    struct multiprecision_integer_type
+    {
+        typedef boost::multiprecision::number
+            <
+                boost::multiprecision::cpp_int_backend
+                    <
+                        Size,
+                        Size,
+                        IntegerType,
+                        check_policy_type::value,
+                        void
+                    >
+            > type;
+    };
+
+    // Meta-function to get the multiprecision signed integer type for
+    // the given size
+    template <unsigned int Size>
+    struct multiprecision_signed_integer_type
+    {
+        typedef typename multiprecision_integer_type
+            <
+                Size, boost::multiprecision::signed_magnitude
+            >::type type;
+    };
+
+    // Meta-function to get the multiprecision unsigned integer type for
+    // the given size
+    template <unsigned int Size>
+    struct multiprecision_unsigned_integer_type
+    {
+        typedef typename multiprecision_integer_type
+            <
+                Size, boost::multiprecision::unsigned_magnitude
+            >::type type;
+    };
+
+    // Define the minimum (and maximum) bit size for the
+    // multiprecision integer needed
+    // If T is the input type and P the promoted type, then the
+    // minimum number of bits for P are (below b stands for the number
+    // of bits of T):
+    // * if T is unsigned and P is unsigned: 2 * b
+    // * if T is signed and P is signed: 2 * b - 1
+    // * if T is unsigned and P is signed: 2 * b + 1
+    typedef typename boost::mpl::if_c
+        <
+            (PromoteUnsignedToUnsigned && is_unsigned),
+            boost::integral_constant
+                <
+                    unsigned int, (2 * CHAR_BIT * sizeof(T))
+                >,
+            boost::integral_constant
+                <
+                    unsigned int,
+                    (2 * CHAR_BIT * sizeof(T) + (is_unsigned ? 1 : -1))
```

Comment:
* `cpp_int_backend<>` expects `unsigned int`, which is why I used it here. For this reason I cannot really use `mpl::size_t` here.
* `mpl::size_t` can be used in other places though. Will make the change when appropriate.
* No, you may not decrease the number of bits by 1 for unsigned numbers. Let me give you an example in binary. Suppose you have 3-bit unsigned numbers: they range from `0` to `111`. If you multiply `111 * 111` you get `110001`, which is well above what you can get by considering 5-bit unsigned numbers (here `5 = 2 * 3 - 1`).
* I am pretty sure that multiprecision numbers increase their bit sizes in chunks. My point of view is that what you describe is the responsibility of the multiprecision number to handle not ours. But yes, we can go over this again later.